pkg/server/observer.go

```diff
@@ -416,52 +458,39 @@ func getFlows(
 		return nil, err
 	}
 
-	log.Debug("filters", zap.String("req", fmt.Sprintf("%+v", req)))
-	log.Debug("whitelist", zap.String("whitelist", fmt.Sprintf("%+v", whitelist)))
-	log.Debug("blacklist", zap.String("blacklist", fmt.Sprintf("%+v", blacklist)))
-
-	ch, stop, err := getBufferCh(ctx, ring, req)
-	if err != nil {
-		return nil, err
-	}
-	reply := make(chan *pb.Flow, 1)
-	go func() {
-		defer close(reply)
-		defer stop()
-
-		for e := range ch {
-			if req.Number != 0 && !req.Follow {
-				i++
-				if i >= req.Number {
-					// stop the channel buffer because we have reached
-					// the number of requested flows.
-					stop()
-					if i > req.Number {
-						// We will 'continue' since 'ch' might have flows and we
-						// need to empty that channel.
-						return
-					}
-				}
-			}
-			if e == nil {
-				continue
-			}
-			flow, ok := e.Event.(*pb.Flow)
-			if !ok || !filters.Apply(whitelist, blacklist, e) {
-				continue
+	idx := ring.LastWriteParallel()
+	reader := container.NewRingReader(ring, idx)
+
+	var flowsCount uint64
+	// We need to find out what the right index is; that is the index with the
+	// oldest entry that is within time range boundaries (if any is defined)
+	// or until we find enough events.
+	// In order to avoid buffering events, we have to rewind first to find the
+	// correct index, then create a new reader that starts from there
+	for i := ring.Len(); i > 0; i, idx = i-1, idx-1 {
+		e := reader.Previous(ctx)
+		if e == nil {
+			continue
+		}
+		_, ok := e.Event.(*pb.Flow)
+		if !ok || !filters.Apply(whitelist, blacklist, e) {
+			continue
+		}
+		flowsCount++
+		if since != nil {
+			ts, err := types.TimestampFromProto(e.GetFlow().GetTime())
+			if err != nil {
+				return nil, err
 			}
-			select {
-			case reply <- flow:
-				// We have sent all expected flows so we can return already
-				if req.Number != 0 && i >= req.Number {
-					return
-				}
-			case <-ctx.Done():
-				return
+			if ts.Before(start) {
+				idx++ // we went backward 1 too far
```

Comment:
This is quite prone to races (since writing happens very fast, concurrently).

If you got back to the oldest entry and will now ++ it to return, it's very likely it's going to be ovrewritten.

I'm not sure that there is a good way to resolve it without read locking elements (which is probably not necessary here). Perhaps we should just buffer more than one element on the end. Maybe we do a % cut-off, like 1,2,3,5% at the end of the reader we assume is for "staging/rewriting" area.

What do you think?

Reply:
I think that this whole code basically is but don't really see it as a problem in this context in the sense that it's no worse that it was.
For sure, I think that we don't want to create lock contention on the ring buffer so your idea of a staging area is probably worth exploring.